ConfigurationBinder handles ISet<>

SteveDunn opened the pull request:
Fixes #66141

SteveDunn committed:
First pass

SteveDunn committed:
Remove commented code

SteveDunn committed:
Support IReadOnlySet. Refactor out method that determines most suitable type for collections that are already specified in the destination.

tarekgh commented on src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs:
```diff
@@ -512,12 +532,16 @@ private static void BindProperty(PropertyInfo property, object instance, IConfig
-            Type type = source.GetType().GetInterface("IEnumerable`1", false)!;
-            Type elementType = type.GenericTypeArguments[0];
-            Type genericType = typeof(List<>).MakeGenericType(elementType);
+            var (collectionType, elementType) = ResolveSuitableCollectionFromSourceCollection(source);
```
Can we add some assert here that we never get non-generic IEnumerable interface. Just to protect against any wrong future use of this method.

SteveDunn replied:
Good idea, but that method has gone since the rebase from main

tarekgh commented:
@SteveDunn could you please resolve the file conflicts? 

CC @maryamariyan

SteveDunn committed:
Handles ISet after rebase. Still to do IReadOnlySet

SteveDunn committed:
Add support back for `IReadOnlySet`

tarekgh commented on src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs:
```diff
@@ -564,6 +593,47 @@ private static Array BindArray(Type type, IEnumerable? source, IConfiguration co
+            {
+                foreach (object? item in source)
+                {
+                    addMethod.Invoke(instance, new[] {item});
```
>new[]  [](http://example.com/codeflow?start=47&length=6)

we always passing the arguments in array of one item. can we just allocate the array once and use it for all calls? This will save many allocations I guess. We can use this array down in this method too.

SteveDunn committed:
PR feedback - pre-allocate just one array instead of many allocations in a loop

SteveDunn commented:
> @SteveDunn could you please resolve the file conflicts?
> 
> CC @maryamariyan

Rebased from main and resolved conflicts.

tarekgh reviewed (approved):
LGTM. Thanks @SteveDunn 

@maryamariyan do you want to have a quick look?

halter73 commented:
The `ArgumentException` when trying to bind to `ISet` shows there's a more general bug in our collection binding logic. This PR fixes `ISet` by special-casing `ISet` before we get to the buggy logic, but we should have never tried to bind to the `ISet` property using a `List` in the first place. We should have always been able to recognize our inability to construct a valid implementation of an interface and skip over it when binding.

It might end up that we test for `ISet` assignability just as this PR already does after fixing the more general bug, but I think there's good chance that rethinking our collection binding logic to avoid assignments that could never work might also change how we test for `ISet.

I would like a fix to `FindOpenGenericInterface` like I suggested in https://github.com/dotnet/runtime/issues/66141#issuecomment-1075743040 so the `SkipsCustomCollection` test below also passes.

```csharp
public interface ICustomCollection<out T> : IEnumerable<T>
{
}
public class MyClassWithCustomCollection
{
    public ICustomCollection<string> CustomCollection { get; set; }
}
[Fact]
public void SkipsCustomCollection()
{
    var configurationBuilder = new ConfigurationBuilder();
    configurationBuilder.AddInMemoryCollection( new Dictionary<string, string>
    {
        ["CustomCollection:0"] = "Yo!",
    });
    var config = configurationBuilder.Build();
    var instance = config.Get<MyClassWithCustomCollection>()!;
    Assert.Null(instance.CustomCollection);
}
```

SteveDunn committed:
All tests pass.

SteveDunn committed:
Tidy

SteveDunn committed:
Skips custom sets

SteveDunn committed:
Strip out unused code

SteveDunn committed:
Tidy up and remove dead code.

SteveDunn committed:
Tidy up tests

SteveDunn committed:
Tidy up comments

SteveDunn commented:
Thanks for the suggestion @halter73 .  I've fixed the issues mentioned, and the example test case you provided is now included. I also added similar cases for custom dictionaries.

halter73 commented on src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs:
```diff
@@ -368,29 +338,33 @@ private static object BindToCollection(Type type, IConfiguration config, BinderO
+                        if (instance != null)
+                        {
+                            bindingPoint.SetValue(instance);
+                        }
```
I think hiding this logic in `AttemptBindToCollectionInterfaces` obfuscates what's really happening.

I checked out your branch because this logic is really complicated and I wanted to make sure my suggestions didn't break any important scenarios. 

Here are the follow up changes I would make to simplify this logic.

https://github.com/dotnet/runtime/compare/fcddddc80204ac920ee23d06f0af84e212792237...halter73/configbinder-iset

While doing this I realized that the more consistent behavior is to throw for null properties with a custom interface type, so I changed the behavior and updated the test.

SteveDunn replied:
Thanks for help. I've copied over your changes.

halter73 reviewed (changes requested):
Upon further review I think we should throw for custom interfaces implementing known collection interface, it just shouldn't be an `ArgumentException` thrown when trying to set a property value, it should be an `InvalidOperationException` thrown while trying to activate the type like we do for other interfaces.

In my last comment I link to a diff to your PR that has the behavior I want as far as that goes. I didn't go ahead and make all the changes I suggested. For example, we still need to limit sets to strings and enums and write tests for that, but please feel free to cherry-pick these changes.

SteveDunn committed:
Copy over @halter73's changes

SteveDunn committed:
Copy over @halter73's changes

SteveDunn committed:
Copy over @halter73's changes

SteveDunn commented on src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs:
```diff
@@ -564,6 +548,52 @@ private static Array BindArray(Type type, IEnumerable? source, IConfiguration co
+        {
+            Type elementType = type.GetGenericArguments()[0];
+
+            Type genericType = typeof(HashSet<>).MakeGenericType(type.GenericTypeArguments[0]);
```
Added restrictions and some tests.

SteveDunn committed:
Added restrictions and some tests for set keys being string/enum only

SteveDunn commented:
> 

I added the above test and got it passing, but there was no way for this to remain green whilst all others remained green.
I then noticed some existing tests that asserts that an exception is thrown for a custom collections deriving from `IEnumerable<>`, e.g. a test called `ThrowsForCustomIEnumerableCollection`:

Do we want to change this behaviour from throwing this exception to just skipping the custom collections (enumerables, sets, dictionaries)?

SteveDunn commented:
This is one of a few PRs laying dormant. Is there anything I need to do to progress them?

halter73 commented:
> Do we want to change this behaviour from throwing this exception to just skipping the custom collections (enumerables, sets, dictionaries)?

No. It's better to stay consistent and throw from all custom collections. That's why I changed all the `SkipCustom...` tests to `ThrowsForCustom...` tests in https://github.com/dotnet/runtime/compare/dotnet:fcddddc...dotnet:dbd9dcd.

halter73 reviewed (approved):
Sorry for the slow turnaround getting back to this. It looks really good. Thanks for sticking with this so long!

I still do want to add a test for a collection without a public `Add` method since we nearly regressed this. I asked for this, but it was tucked away in a longer comment.

> We should add a test that binds to an explicitly implemented `ICollection<string>` to verify we don't regress this in the future.

eerhardt commented:
> I still do want to add a test for a collection without a public Add method since we nearly regressed this.

@SteveDunn - I think this is the last feedback to address on this PR, and then it can be merged.

eerhardt commented on src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs:
```diff
@@ -631,16 +639,38 @@ private static Array BindArray(Type type, IEnumerable? source, IConfiguration co
+
+            Type genericTypeDefinition = type.GetGenericTypeDefinition();
+            return genericTypeDefinition == typeof(ISet<>)
+#if NET5_0_OR_GREATER
```
```suggestion
#if NETCOREAPP
```
See Commit 1 of https://github.com/dotnet/runtime/pull/65847

eerhardt commented on src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs:
```diff
@@ -368,34 +312,42 @@ private static object BindToCollection(Type type, IConfiguration config, BinderO
+                        Type typeOfKey = type.GenericTypeArguments[0];
+                        Type typeOfValue = type.GenericTypeArguments[1];
+                        // Overwrite type in case it was a IReadOnlyDictionary<>. We still want to be able to bind items.
+                        // REVIEW: What about settable IReadOnlyDictionary<> instances with an initial value?
+                        // I think we should consider preferring copying like we do for all other collection interfaces.
```
What is the plan to address this comment?

SteveDunn replied:
The last few commits is a first pass attempt at addressing this. Dictionary _interfaces_ are cloned and bound. 
_Concrete_ dictionaries are _not_ cloned, as this would be a breaking change to existing behaviour.
I've marked this as resolved, but I imagine there to be a few more comments re. the implementation.

eerhardt replied:
Can we remove the comment?

eerhardt reviewed (commented):
I guess I had 2 more comments.

SteveDunn committed:
PR feedback re. preprocessor directive

SteveDunn committed:
Started on instantiated dictionary feedback

SteveDunn committed:
First pass at cloning dictionary<,>

SteveDunn committed:
Can now bind (overwrite) to an existing concrete dictionary

SteveDunn committed:
Refactor out common code for binding to existing dictionary. Added tests.

SteveDunn committed:
Comments and more tests

SteveDunn committed:
Add explicit (separate) test for binding to non-instantiated IReadOnlyDictionary

xakep139 commented on src/libraries/Microsoft.Extensions.Configuration.Binder/tests/ConfigurationBinderTests.cs:
```diff
@@ -75,6 +121,87 @@ public override string Virtual
+
+            public IReadOnlyDictionary<string, int> Dictionary { get; set; } =
+                _existingDictionary;
+
```
```suggestion
```

xakep139 commented on src/libraries/Microsoft.Extensions.Configuration.Binder/tests/ConfigurationBinderTests.cs:
```diff
@@ -49,7 +56,46 @@ public string ReadOnly
+                };
+
             public IEnumerable<string> NonInstantiatedIEnumerable { get; set; } = null!;
+            public ISet<string> InstantiatedISet { get; set; } = new HashSet<string>();
```
Could you please also add two more test cases with concrete types (instead of `ISet<T>`) - `HashSet<T>` and `SortedSet<T>`?

SteveDunn committed:
PR feedback - add tests

SteveDunn commented:
I think everything has now been addressed for this. And I also just rebased.

eerhardt commented on src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs:
```diff
@@ -719,16 +801,38 @@ private static Array BindArray(Type type, IEnumerable? source, IConfiguration co
 
             Type genericTypeDefinition = type.GetGenericTypeDefinition();
             return genericTypeDefinition == typeof(IEnumerable<>)
+                || genericTypeDefinition == typeof(ICollection<>)
```
I don't think this fits in with the name of the method: `IsArrayCompatibleReadOnlyInterface` - `ICollection<>` is not "read-only".

SteveDunn replied:
I think the method is answering the question 'Should the type be treated as immutable?" - e.g.  if it's one of these types, then return `true` to say **not** to append to the source, but clone it to a new collection.
I've renamed it as such, although I do think it could be clearer. Do you have any thoughts?

eerhardt replied:
Why do we need to check for `ICollection<>` in both places?

SteveDunn replied:
I think the second check for `ICollection<>` is looking for an `Add` method to determine whether to set (add) members of the collection, or whether to bind to properties (line 358).

Do you have any thoughts on how this could be made clearer? To be perfectly honest, I do find this method rather confusing!

eerhardt commented on src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs:
```diff
@@ -367,6 +298,37 @@ private static object BindToCollection(Type type, IConfiguration config, BinderO
                     return;
                 }
 
+                // for sets and read-only set interfaces, we concatenate on to what is already there
```
This comment isn't true, right? `BindSet` creates a new instance.

SteveDunn replied:
Well spotted - thanks!

SteveDunn replied:
I also updated the comment re. dictionaries.

eerhardt commented on src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs:
```diff
@@ -376,34 +338,43 @@ private static object BindToCollection(Type type, IConfiguration config, BinderO
+                // At this point we know that we have a non-null bindingPoint.Value, we just have to populate the items
+                // using the IDictionary<> or ICollection<> interfaces, or properties using reflection.
+                Type? dictionaryInterface = FindOpenGenericInterface(typeof(IDictionary<,>), type);
+                Type? dictionaryInterface2 = FindOpenGenericInterface(typeof(IReadOnlyDictionary<,>), type);
```
1. `dictionaryInterface2` doesn't seem like a great variable name.
2. Do we need to check for `IReadOnlyDictionary` unconditionally, even if we already found `IDictionary`?

SteveDunn replied:
Removed the latter check

SteveDunn committed:
PR feedback: introdcue variable for `typeIsADictionaryInterface'

SteveDunn commented on src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs:
```diff
@@ -376,34 +338,43 @@ private static object BindToCollection(Type type, IConfiguration config, BinderO
+                    // on a new instance of the interface over populating the existing instance implementing the interface.
+                    // This has already been done, so there's not need to check again. For dictionaries, we fill the existing
+                    // instance if there is one (which hasn't happened yet), and only create a new instance if necessary.
+                    if (TypeIsADictionaryInterface(type))
```
Well spotted. This method is crying out for a bit of refactoring! I'm not sure at what point we'd do that though; maybe as a completely separate Issue unrelated to any change in behaviour?
You are correct though, so I removed the other check (and added another test to verify).

SteveDunn committed:
PR feedback

SteveDunn commented:
Might it be possible for this be closed soon? I'm in a situation where I have a few PRs that have been ready to close for some time, and I'm needing to periodically rebase. @eerhardt @halter73 @tarekgh

ericstj commented:
@tarekgh it looks like you signed off here - can you have a look and see what's needed to get this in?

tarekgh commented:
I'll take one last look and then merge. 

@SteveDunn could you please resolve the conflicts one last time? Sorry for the delay!

SteveDunn committed:
Fix rebase issues

SteveDunn commented:
> I'll take one last look and then merge.
> 
> @SteveDunn could you please resolve the conflicts one last time? Sorry for the delay!

No worries at all, you must be busy (if you need an additional member of the team, I'd love to join!)

tarekgh commented:
The failure in Windows x64 Debug run is unrelated and tracked by https://github.com/dotnet/runtime/issues/72428
